ksqldb-common/src/main/java/io/confluent/ksql/util/PushOffsetVector.java

```diff
@@ -18,31 +18,79 @@
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonIgnore;
 import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
 import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
+import java.util.Collections;
+import java.util.HashMap;
 import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
 
 /**
  * OffsetVector for push query continuation tokens.
  */
 public class PushOffsetVector implements OffsetVector {
 
-  private final List<Long> offsets;
+  private final AtomicReference<List<Long>> offsets = new AtomicReference<>();
 
   @SuppressFBWarnings(value = "EI_EXPOSE_REP2")
   @JsonCreator
   public PushOffsetVector(final @JsonProperty(value = "o") List<Long> offsets) {
-    this.offsets = offsets;
+    this.offsets.set(ImmutableList.copyOf(offsets));
+  }
+
+  public PushOffsetVector() {
```

Comment:
I agree that there's some overlap here.  We basically started off by saying that we would try to create a common interface and then we could use different implementations since I might want to store mine densely and serialize them differently.  In retrospect, after looking at them, they're similar enough that they should probably be merged at some point.

I added some of these thoughts to the documentation. 

Also, there are more issues with the "consistent pull query then read" in that the offsets are tracking different things.  For example: `CREATE table RATINGS_TABLE as SELECT * from SOURCE; SELECT * FROM RATINGS_TABLE WHERE COL > 10;`. It's my understanding that the offsets stored in the materialization and used for consistency are going to be the offsets of `SOURCE` whereas with SPQ, the offsets used for continuation are those of `RATINGS_TABLE`.  We can still potentially figure out some way to do this, such as just ask streams what are the current offsets written to `RATINGS_TABLE` before a pull query, but it's not going to be quite as straightforward as we were abstractly thinking.

Reply:
That's correct. This is the main technical work we'll have to tackle before we can support "query and subscribe".